### PR TITLE
Message content for Fail assertion

### PR DESCRIPTION
**Pull request by M6AI**
## Bug / Requirement Description
Fail assertion only carries content in the form of description which can become unaesthetic in case of longer explanation for the intended failure.

## Solution description
Added message attribute and associated rendering to fail assertion in the likeness of log.

## Checklist:
- [ ] Test
- [ ] Example (both test_plan.py and .rst)
- [ ] Documentation (API)
- [ ] News fragment present for release notes
- [ ] MS info leakage check
- [ ] For new driver: driver index page
- [ ] For new assertion: ui/pdf/std renderers, documentation
- [ ] For new cmdline arg: documentation

### CODE REVIEW

**Comment by M6AI** [comment]
Credit to John, Adzagogo. PR still misses newsfragment and a few docstrings, will apply soon.

File: testplan/testing/multitest/entries/stdout/assertions.py [hunk 1]
```diff
@@ -463,10 +463,22 @@ def get_assertion_details(self, entry):
...
 @registry.bind(assertions.Fail)
 class FailRenderer(AssertionRenderer):
-    def get_header_text(self, entry):
-        """Always return text in red as it is an explicit failure."""
-        return Color.red(entry.description)
+    def get_header(self, entry):
+        if entry.description:
+            return Color.red(entry.description)
+        elif isinstance(entry.message, str):
+            return Color.red(str(entry.message))
```
Comment (M6AI) [inline testplan/testing/multitest/entries/stdout/assertions.py:470]: Not sure if we need this.

File: testplan/testing/multitest/entries/stdout/assertions.py [hunk 2]
```diff
@@ -463,10 +463,22 @@ def get_assertion_details(self, entry):
...
+            return self.get_default_header(entry)
 
+    def get_assertion_details(self, entry):
+        if isinstance(entry.message, str):
+            if entry.description:
+                return Color.red(str(entry.message))
+            else:
+                return None
+        else:
+            return pprint.pformat(entry.message)
```
Comment (M6AI) [inline testplan/testing/multitest/entries/stdout/assertions.py:481]: Missing import for pprint.

File: testplan/testing/multitest/entries/assertions.py [hunk 3]
```diff
@@ -122,6 +122,24 @@ def evaluate(self):
...
+        if isinstance(message, str):
+            self.message = message
+        elif isinstance(message, bytes):
+            self.message = message.decode()
+        else:
+            self.message = pprint.pformat(message)
+
+        if not description:
+            description = next((l for l in self.message.split("\n") if l), "")
+        if len(description) > 80:
```
Comment (M6AI) [inline testplan/testing/multitest/entries/assertions.py:137]: Needs indentation.

File: testplan/testing/multitest/entries/stdout/assertions.py (hunk 1 above)
Comment (Pyifan) [reply testplan/testing/multitest/entries/stdout/assertions.py:471]: agree, i think we always have a description

File: testplan/testing/multitest/entries/assertions.py [hunk 4]
```diff
@@ -122,6 +123,25 @@ def evaluate(self):
 
 
 class Fail(Assertion):
+    def __init__(
+        self, description=None, category=None, flag=None, message=None
```
Comment (Pyifan) [inline testplan/testing/multitest/entries/assertions.py:127]: To achieve the same behavior with result.log, shall we have 'message' as the first and required parameter?

File: testplan/testing/multitest/result.py [hunk 5]
```diff
@@ -1608,7 +1608,13 @@ def log_code(self, code, language="python", description=None):
         return entry
 
     @assertion
-    def fail(self, description, category=None, flag=None):
+    def fail(
```
Comment (Pyifan) [inline testplan/testing/multitest/result.py:1611]: We usually use result.log like this:
result.log("This is a log assertion")
result.log(cmd_output, description="execute {cmd}")

I would expect result.fail to exhibits same usage and behavior. Thus shall we change the order of the parameters, and make it consistent with result.log?

**Pull request merged**